[ENH] Implement Notification Icon with Badge for Warnings and Its Test

Tusharjamdade opened the pull request:
<!--- Until this PR is ready for review, you can include [WIP] in the title, or create a draft PR. -->


<!---
Below is a suggested pull request template. Feel free to add more details you feel are relevant/necessary.

For more info on the Neurobagel PR process and other contributing guidelines, see https://neurobagel.org/contributing/CONTRIBUTING/.
-->

<!-- 
Please indicate after the # which issue you're closing with this PR, if applicable.
If the PR closes multiple issues, include "closes" before each one is listed.
You can also link to other issues if necessary, e.g. "See also #1234".

https://help.github.com/articles/closing-issues-using-keywords
-->
- Closes #393 

<!-- 
Please give a brief overview of what has changed or been added in the PR.
This can include anything specific the maintainers should be looking for when they review the PR.
-->
Changes proposed in this pull request:

- Implemented a notification badge for non-critical warnings
- Additionally, added a test for the notification badge


<!-- To be checked off by reviewers -->
## Checklist
_This section is for the PR reviewer_

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[FIX]`, `[REF]`, `[TST]`, `[CI]`, `[MNT]`, `[INF]`, `[MODEL]`, `[DOC]`) _(see our [Contributing Guidelines](https://neurobagel.org/contributing/CONTRIBUTING#pull-request-guidelines) for more info)_
- [x] PR has a label for the release changelog or `skip-release` (to be applied by maintainers only)
- [x] PR links to GitHub issue with mention `Closes #XXXX`
- [x] Tests pass
- [x] Checks pass
- [ ] If the PR changes the participant-level and/or the dataset-level result file, the [`query-tool-results` files](https://github.com/neurobagel/neurobagel_examples/tree/main/query-tool-results) in the  [neurobagel_examples repo](https://github.com/neurobagel/neurobagel_examples/) have been regenerated

For new features:
- [x] Tests have been added

For bug fixes:
- [ ] There is at least one test that would fail under the original bug conditions.

## Summary by Sourcery

Add a notification icon with a badge to display non-critical warnings to the user.

New Features:
- Display non-critical warnings as notifications in a popover.

Tests:
- Add tests for the notification icon and badge.

Tusharjamdade committed:
Implement notification icon with badge for warnings and its test

rmanaem committed:
Merge branch 'main' into warnings

rmanaem commented on src/components/Navbar.tsx:
```diff
@@ -8,30 +8,55 @@ import {
   const openAccountMenu = Boolean(anchorEl);
 
   const { user, logout } = useAuth0();
+  const [anchorMailEl, setAnchorMailEl] = useState<null | HTMLElement>(null);
```
```suggestion
  const [anchorNotifEl, setAnchorNotifEl] = useState<null | HTMLElement>(null);
```

Since we switched out the mail icon for notification, it's good to update the relevant variables.

rmanaem commented on src/components/Navbar.tsx:
```diff
@@ -8,30 +8,55 @@ import {
 
   const { user, logout } = useAuth0();
+  const [anchorMailEl, setAnchorMailEl] = useState<null | HTMLElement>(null);
+  const openMailMenu = Boolean(anchorMailEl);
```
```suggestion
  const openNotifMenu = Boolean(anchorNotifEl);
```

rmanaem commented on src/components/Navbar.tsx:
```diff
@@ -8,30 +8,55 @@ import {
 
+  const handleMailClick = (event: React.MouseEvent<HTMLElement>) => {
+    setAnchorMailEl(event.currentTarget);
+  };
```
```suggestion
  const handleNotifClick = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorNotifEl(event.currentTarget);
  };
```

rmanaem commented on src/components/Navbar.tsx:
```diff
@@ -8,30 +8,55 @@ import {
+  };
+
+  const handleMailClose = () => {
+    setAnchorMailEl(null);
```
```suggestion
  const handleNotifClose = () => {
    setAnchorNotifEl(null);
```

rmanaem commented on src/components/Navbar.tsx:
```diff
@@ -56,9 +81,83 @@ function Navbar({ isLoggedIn, onLogin }: { isLoggedIn: boolean; onLogin: () => v
             </IconButton>
           </Tooltip>
+          <Tooltip title="Notifications">
+            <IconButton onClick={handleMailClick}>
```
```suggestion
            <IconButton onClick={handleNotifClick}>
```

rmanaem commented on src/components/Navbar.tsx:
```diff
@@ -56,9 +81,83 @@ function Navbar({ isLoggedIn, onLogin }: { isLoggedIn: boolean; onLogin: () => v
+          <Popover
+            open={openMailMenu}
+            anchorEl={anchorMailEl}
+            onClose={handleMailClose}
```
```suggestion
            open={openNotifMenu}
            anchorEl={anchorNotifEl}
            onClose={handleNotifClose}
```

rmanaem commented on src/components/Navbar.tsx:
```diff
@@ -56,9 +81,83 @@ function Navbar({ isLoggedIn, onLogin }: { isLoggedIn: boolean; onLogin: () => v
+              vertical: 'top',
+              horizontal: 'right',
+            }}
+            PaperProps={{
```
I believe the `PaperProps` is deprecated see the relevant documentation section [here](https://v5.mui.com/material-ui/api/popover/#popover-prop-PaperProps)

rmanaem commented on src/components/Navbar.tsx:
```diff
@@ -56,9 +81,83 @@ function Navbar({ isLoggedIn, onLogin }: { isLoggedIn: boolean; onLogin: () => v
+                      primary={
+                        <Typography variant="subtitle2" className="font-bold text-[#E65100]">
+                          {/* Add the appropriate title for the notification if any */}
+                        </Typography>
```
I think we can use the type of notification e.g., warning, error, info for the primary content.

rmanaem commented on src/components/Navbar.tsx:
```diff
@@ -56,9 +81,83 @@ function Navbar({ isLoggedIn, onLogin }: { isLoggedIn: boolean; onLogin: () => v
+                        edge="end"
+                        aria-label="delete"
+                        className="text-[#FF5722] hover:bg-[#FF572220]"
+                      />
```
You may have forgotten to add the delete icon and implement the logic for deleting a list item.

rmanaem commented on src/components/Navbar.tsx:
```diff
@@ -56,9 +81,83 @@ function Navbar({ isLoggedIn, onLogin }: { isLoggedIn: boolean; onLogin: () => v
+                  <ListItem
+                    key={notification}
+                    alignItems="flex-start"
+                    className="mb-2 rounded-md border-l-4 border-[#FF9800] bg-[#FFF3E0] transition-colors duration-200 hover:bg-gray-200"
```
The issue with this implementation is it doesn't take into account the type of notification and colors them all the same see other comments for more context

rmanaem commented on cypress/component/Navbar.cy.tsx:
```diff
@@ -24,5 +31,7 @@ describe('Navbar', () => {
       .eq(1)
       .should('have.attr', 'href', 'https://github.com/neurobagel/query-tool/');
+    // Test for Notification Icon and Badge
+    cy.get("[data-cy='navbar'] button").eq(1).find('svg').should('be.visible'); // Check if the notification icon exists
```
It's good to expand this test to include some assertions about the contents of the popover

rmanaem reviewed (changes requested):
Hi @Tusharjamdade,

Congratulations on your first contribution to Neurobagel! 🎉 The changes look great overall. However, there are a few areas that need adjustments:

- Currently, the implementation doesn't support deleting notifications, either individually or all at once.
- The way warnings are passed to the Navbar component is too specific, limiting reusability.
- Some of the e2e and component tests are failing due to these changes.

To address these, we can:

- Implement delete functionality for each notification and a "clear all notifications" feature for the popover.
- Refactor the logic for passing notifications to the Navbar component to make it more generic. This will allow us to pass different types of notifications (info, errors) to the Navbar.
  -  For that the type of notification so that the appropriate title and color can be set in the ListItemText.
  - Once the logic is updated, we can also display info notifications in the popover.
- Update the tests to reflect these changes.

I understand this might be overwhelming for your first PR and contribution. Please don't hesitate to reach out with any questions, or let us know if you would prefer one of the maintainers to take over this PR.

Tusharjamdade commented:
> Hi @Tusharjamdade,
> 
> Congratulations on your first contribution to Neurobagel! 🎉 The changes look great overall. However, there are a few areas that need adjustments:
> 
> * Currently, the implementation doesn't support deleting notifications, either individually or all at once.
> * The way warnings are passed to the Navbar component is too specific, limiting reusability.
> * Some of the e2e and component tests are failing due to these changes.
> 
> To address these, we can:
> 
> * Implement delete functionality for each notification and a "clear all notifications" feature for the popover.
> * Refactor the logic for passing notifications to the Navbar component to make it more generic. This will allow us to pass different types of notifications (info, errors) to the Navbar.
>   
>   * For that the type of notification so that the appropriate title and color can be set in the ListItemText.
>   * Once the logic is updated, we can also display info notifications in the popover.
> * Update the tests to reflect these changes.
> 
> I understand this might be overwhelming for your first PR and contribution. Please don't hesitate to reach out with any questions, or let us know if you would prefer one of the maintainers to take over this PR.

Thank you for the review, @rmanaem.

I appreciate the detailed feedback and suggestions.
I will work on the details shared and will reach out if I need any help!!

Tusharjamdade commented:
> Hi @Tusharjamdade,
> 
> Congratulations on your first contribution to Neurobagel! 🎉 The changes look great overall. However, there are a few areas that need adjustments:
> 
> * Currently, the implementation doesn't support deleting notifications, either individually or all at once.
> * The way warnings are passed to the Navbar component is too specific, limiting reusability.
> * Some of the e2e and component tests are failing due to these changes.
> 
> To address these, we can:
> 
> * Implement delete functionality for each notification and a "clear all notifications" feature for the popover.
> * Refactor the logic for passing notifications to the Navbar component to make it more generic. This will allow us to pass different types of notifications (info, errors) to the Navbar.
>   
>   * For that the type of notification so that the appropriate title and color can be set in the ListItemText.
>   * Once the logic is updated, we can also display info notifications in the popover.
> * Update the tests to reflect these changes.
> 
> I understand this might be overwhelming for your first PR and contribution. Please don't hesitate to reach out with any questions, or let us know if you would prefer one of the maintainers to take over this PR.

Hi @rmanaem 
```tsx
export type Notification = {
  id: number;
  type: 'info' | 'warning';
  message: string;
};

const [notifications, setNotifications] = useState<Notification[]>([]);

<Navbar
  isLoggedIn={isAuthenticated}
  onLogin={() => setOpenAuthDialog(true)}
  notifications={notifications}
  setNotifications={setNotifications}
/>

// Navbar component
function Navbar({
  isLoggedIn,
  onLogin,
  notifications,
  setNotifications
}: {
  isLoggedIn: boolean;
  onLogin: () => void;
  notifications: Notification[];
  setNotifications: React.Dispatch<React.SetStateAction<Notification[]>>;
}) 
```
This is an overview of the implementation I have used. I have stored all notifications based on their types (warning, info) and messages, assigning each one a unique identifier (which will be used to delete the notification when the delete icon is clicked).

I passed the notifications and setNotifications to update the notifications (when the user clicks the delete button or the clear all button). Is this approach acceptable, or do I need to work on something else?

rmanaem commented:
Hi @Tusharjamdade,
That's a good approach. Could you push your changes so I can take a look at them in context?

Tusharjamdade commented:
> Hi @Tusharjamdade, That's a good approach. Could you push your changes so I can take a look at them in context?

Hi @rmanaem,

I am encountering a strange error

![image](https://github.com/user-attachments/assets/ba7daa2e-1690-439e-b0a1-26a6e1eeff46)

It checks whether the `enqueueSnackbar` contains, for example, `NoAssessmentNode`, but this value was never passed to the `enqueueSnackbar` function as a warning in the `App.tsx` file or any other file. As a result, the test fails because it is looking for `NoAssessmentNode`, which does not exist. I modified the `enqueueSnackbar` function and added all the warnings to the notification, but since the value `NoAssessmentNode` was never passed to the notification, the test still fails. 
Am I missing something?

```tsx
it('Shows warning for node that failed Assessment tool option request', () => {
    cy.get('.notistack-SnackbarContainer')
      .find('.notistack-MuiContent-warning')
      .should('contain', 'NoAssessmentNode');
  });
  it('Shows warning for node that failed Diagnosis option request', () => {
    cy.get('.notistack-SnackbarContainer')
      .find('.notistack-MuiContent-warning')
      .should('contain', 'NoDiagnosisNode');
  });
  it('Shows warning for node that failed Pipeline option request', () => {
    cy.get('.notistack-SnackbarContainer')
      .find('.notistack-MuiContent-warning')
      .should('contain', 'NoPipelineNode');
  });
  it('Shows warning for node that failed Pipeline version option request', () => {
    cy.get('[data-cy="close-auth-dialog-button"]').click();
    cy.get('[data-cy="Pipeline name-categorical-field"]').type('fmri{downarrow}{enter}');
    cy.wait('@getPipelineVersionsOptions');
    cy.get('.notistack-SnackbarContainer')
      .find('.notistack-MuiContent-warning')
      .should('contain', 'NoPipelineVersionNode');
  });
```

Tusharjamdade committed:
Improved notification functionality and added e2e & component tests

Tusharjamdade commented:
> > Hi @Tusharjamdade, That's a good approach. Could you push your changes so I can take a look at them in context?
> 
> Hi @rmanaem,
> 
> I am encountering a strange error
> 
> ![image](https://private-user-images.githubusercontent.com/129779656/404546434-ba7daa2e-1690-439e-b0a1-26a6e1eeff46.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzcyMzIzNjAsIm5iZiI6MTczNzIzMjA2MCwicGF0aCI6Ii8xMjk3Nzk2NTYvNDA0NTQ2NDM0LWJhN2RhYTJlLTE2OTAtNDM5ZS1iMGExLTI2YTZlMWVlZmY0Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDExOFQyMDI3NDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03YTcxOGZlMzU1Y2Y2YThkNTAxY2NkNjU4OTU4OGJkOWI1N2ExNzE0OTY4MTc3ZWQ5OTJjM2U4YzZmODRiYjc0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.JKbX_N0RnpblsebqxFdSyROcAcf8ceplBB7WVT_frfk)
> 
> It checks whether the `enqueueSnackbar` contains, for example, `NoAssessmentNode`, but this value was never passed to the `enqueueSnackbar` function as a warning in the `App.tsx` file or any other file. As a result, the test fails because it is looking for `NoAssessmentNode`, which does not exist. I modified the `enqueueSnackbar` function and added all the warnings to the notification, but since the value `NoAssessmentNode` was never passed to the notification, the test still fails. Am I missing something?
> 
> ```tsx
> it('Shows warning for node that failed Assessment tool option request', () => {
>     cy.get('.notistack-SnackbarContainer')
>       .find('.notistack-MuiContent-warning')
>       .should('contain', 'NoAssessmentNode');
>   });
>   it('Shows warning for node that failed Diagnosis option request', () => {
>     cy.get('.notistack-SnackbarContainer')
>       .find('.notistack-MuiContent-warning')
>       .should('contain', 'NoDiagnosisNode');
>   });
>   it('Shows warning for node that failed Pipeline option request', () => {
>     cy.get('.notistack-SnackbarContainer')
>       .find('.notistack-MuiContent-warning')
>       .should('contain', 'NoPipelineNode');
>   });
>   it('Shows warning for node that failed Pipeline version option request', () => {
>     cy.get('[data-cy="close-auth-dialog-button"]').click();
>     cy.get('[data-cy="Pipeline name-categorical-field"]').type('fmri{downarrow}{enter}');
>     cy.wait('@getPipelineVersionsOptions');
>     cy.get('.notistack-SnackbarContainer')
>       .find('.notistack-MuiContent-warning')
>       .should('contain', 'NoPipelineVersionNode');
>   });
> ```

I have resolved this issue and updated the changes. Could you please review them?

Tusharjamdade committed:
revert changes

rmanaem committed:
Merge branch 'main' into warnings

rmanaem commented on src/App.tsx:
```diff
@@ -116,26 +119,38 @@ function App() {
+            setNotifications((prev) => [
+              ...prev,
+              {
+                id: prev.length + 1,
```
I don't think this is a safe way of setting the `id`, let's use a safer approach like UUID.

rmanaem commented on cypress/component/Navbar.cy.tsx:
```diff
@@ -24,5 +39,25 @@ describe('Navbar', () => {
+      });
+    });
+    cy.get('.MuiPopover-paper').trigger('keydown', { key: 'Escape' });
+    cy.get('.MuiPopover-paper').should('not.exist');
```
Very nice!

rmanaem commented on src/components/Navbar.tsx:
```diff
@@ -56,9 +92,111 @@ function Navbar({ isLoggedIn, onLogin }: { isLoggedIn: boolean; onLogin: () => v
+                    fontWeight: 'bold',
+                    '&:hover': {
+                      backgroundColor: '#f50057',
+                      color: '#fff',
```
Is there a reason you're using the secondary color and a custom color for the background?

Tusharjamdade replied:
I will remove the secondary color as it is being overridden

rmanaem replied:
I'd also suggest sticking with the existing color theme unless there is an issue.

rmanaem commented on cypress/e2e/APIRequests.cy.ts:
```diff
@@ -505,9 +540,11 @@ describe('Partially successful API query requests', () => {
-    cy.get('.notistack-SnackbarContainer')
-      .find('.notistack-MuiContent-warning')
-      .should('contain', 'DidNotWorkNode');
+    cy.get("[data-cy='navbar']").find('svg[data-testid="NotificationsIcon"]').should('exist');
```
Feel free to add a `data-cy` to the notification button so you don't have to access it through the navbar every time.

rmanaem commented on cypress/component/Navbar.cy.tsx:
```diff
@@ -24,5 +39,25 @@ describe('Navbar', () => {
+      });
+    });
+    cy.get('.MuiPopover-paper').trigger('keydown', { key: 'Escape' });
+    cy.get('.MuiPopover-paper').should('not.exist');
```
I suggest adding some assertions for the newly implemented individual notification delete and clear all features.

Tusharjamdade replied:
I tried implementing the delete and clear all functionalities, but I passed a default notification list. As a result, I was unable to delete notifications, and the tests were failing. I will find a way to implement this.

rmanaem replied:
I see, I think the issue there is that inside the actual component, we're using the `setNotification` to update the `notifications` but inside the test, we mock that one as well.

rmanaem reviewed (changes requested):
Thanks for the great work @Tusharjamdade! The changes look great and the tests are now passing!
I left a few comments, please take a look and feel free to ping me with any questions.

Tusharjamdade committed:
add tests for deleting and clearing all functionality

Tusharjamdade commented:
Hi @rmanaem, 
I have implemented the suggested changes. Could you please review them?

rmanaem committed:
Merge branch 'main' into warnings

rmanaem commented on cypress/component/Navbar.cy.tsx:
```diff
@@ -1,13 +1,28 @@
+import { Notification } from '../../src/utils/types';
 
 describe('Navbar', () => {
+  const mockNotifications: Notification[] = [
+    { id: '1', type: 'info', message: 'This is an info notification.' },
+    { id: '2', type: 'warning', message: 'This is a warning notification.' },
+  ];
+
+  function NavbarWrapper() {
+    const [notifications, setNotifications] = useState(mockNotifications);
+
+    return (
+      <Navbar
+        isLoggedIn
+        onLogin={() => {}}
+        notifications={notifications}
+        setNotifications={setNotifications}
+      />
+    );
+  }
+
```
Very nice!

rmanaem reviewed (approved):
Thanks for making the changes @Tusharjamdade! This is ready to merge 🧑‍🍳
